BUG: Do not rely on undefined behaviour to cast from float to datetime

siddhesh opened the pull request:
A cast from float to signed integer is only valid for the range of
integer, which means that a cast of a NaN to int is undefined.  On x86
cvttd2si conveniently seems to put the mantissa into the long (I
haven't done an exhaustive check to see if that is what it does, but
seems that way), which works for conversion from NaN to NaT.  On
aarch64 however, the corresponding fcvtzs sets the value to 0 for all
undefined ranges, thus breaking the cast.

This fix splits out the float -> integer conversions, explicitly
checks for NaN and sets the target value as NaT, which should make the
behaviour well defined for NaN to NaT translations.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

siddhesh commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -1119,7 +1199,14 @@ static void
+        /* NaN to integer is undefined, so only support cases where NaN for the
+         * target format has a special meaning, i.e. the date/time NaT. */
+        if (npy_isnan(f))
+            *op++ = (@totype@)NPY_DATETIME_NAT;
```
Fixed, updated patch to make the new pattern specific to datetime.  Thanks for your review!

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -1075,12 +1075,12 @@ TIMEDELTA_setitem(PyObject *op, void *ov, void *vap)
  * #TOTYPE = BYTE, UBYTE, SHORT, USHORT, INT, UINT, LONG, ULONG,
- *           LONGLONG, ULONGLONG, FLOAT, DOUBLE, LONGDOUBLE, DATETIME,
+ *           LONGLONG, ULONGLONG, FLOAT, DOUBLE, LONGDOUBLE,
  *           TIMEDELTA#
```
I think timedelta should be handled alongside datetime

eric-wieser commented on numpy/core/tests/test_datetime.py:
```diff
@@ -483,6 +483,9 @@ def test_datetime_nat_casting(self):
         assert_equal(np.datetime64(a, '[W]'), np.datetime64('NaT', '[W]'))
 
+        # NaN -> NaT
+        assert_equal(np.array(np.nan).astype('M8[ns]'), np.datetime64('NaT'))
```
Can you add a test for `m8[ns]` / `np.timedelta64` too?

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -1127,6 +1127,88 @@ static void
+        @fromtype@ f = *ip++;
+
+        /* Avoid undefined behaviour for NaN -> NaT */
+        if (npy_isnan(f)) {
```
An alternative trick you could use here is add a `supports_nat` to the loop template, and then use `#if @supports_nat@` here, which would make the patch shorter.

siddhesh replied:
Thanks for the tip, I've used this.

eric-wieser reviewed (approved):
Looks good to me, will wait to confirm that another maintainer is ok with this loop-merging approach

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -1119,7 +1129,15 @@ static void
 
     while (n--) {
-        *op++ = (@totype@)*ip;
+        @fromtype@ f = *ip++;
```
This is a bug, this line should read `*ip` - the `++` happens on line 1141

siddhesh replied:
Fixed and added more tests to cover this.

eric-wieser reviewed (changes requested):
Typo

siddhesh committed:
BUG: Do not rely on undefined behaviour to cast from float to datetime

A cast from float to signed integer is only valid for the range of
integer, which means that a cast of a NaN to int is undefined.  On x86
cvttd2si conveniently seems to put the mantissa into the long (I
haven't done an exhaustive check to see if that is what it does, but
seems that way), which works for conversion from NaN to NaT.  On
aarch64 however, the corresponding fcvtzs sets the value to 0 for all
undefined ranges, thus breaking the cast.

This explicitly checks for NaN for conversions from floating point
types to DATETIME and TIMEDELTA and sets the target value as NaT,
which should make the behaviour well defined for NaN to NaT
translations.

siddhesh commented:
Ugh, sorry about that.  I've fixed it and also enhanced the test cases to catch this and many other combinations of bugs.  In particular the test now checks for arrays of single, double and long double FP as well as single, double and long double complex FP conversions to datetime and timedelta.

hameerabbasi reviewed (approved):
LGTM.

mattip commented:
Thanks @siddhesh
